Source/player.cpp

```diff
@@ -2665,30 +2662,26 @@ void AddPlrExperience(int pnum, int lvl, int exp)
 	}
 
 	// Adjust xp based on difference in level between player and monster
-	exp = std::max(static_cast<int>(exp * (1 + (lvl - player._pLevel) / 10.0)), 0);
+	uint32_t clampedExp = std::max(static_cast<int>(exp * (1 + (lvl - player._pLevel) / 10.0)), 0);
 
 	// Prevent power leveling
 	if (gbIsMultiplayer) {
-		int powerLvlCap = player._pLevel < 0 ? 0 : player._pLevel;
-		if (powerLvlCap >= 50) {
-			powerLvlCap = 50;
-		}
-		// cap to 1/20 of current levels xp
-		if (exp >= ExpLvlsTbl[powerLvlCap] / 20) {
-			exp = ExpLvlsTbl[powerLvlCap] / 20;
-		}
-		// cap to 200 * current level
-		int expCap = 200 * powerLvlCap;
-		if (exp >= expCap) {
-			exp = expCap;
-		}
-	}
+		const uint32_t clampedPlayerLevel = clamp(static_cast<int>(player._pLevel), 0, 50);
 
-	player._pExperience += exp;
-	if ((DWORD)player._pExperience > MAXEXP) {
-		player._pExperience = MAXEXP;
+		// for low level characters experience gain is capped to 1/20 of current levels xp
+		// for high level characters experience gain is capped to 200 * current level - this is a smaller value than 1/20 of the exp needed for the next level after level 5.
+		clampedExp = std::min({ clampedExp, /* level 0-5: */ ExpLvlsTbl[clampedPlayerLevel] / 20U, /* level 6-50: */ 200U * clampedPlayerLevel });
 	}
 
+	constexpr uint32_t MaxExperience = 2000000000U;
+
+	const uint32_t newExperienceValue = player._pExperience + clampedExp;
+	if (newExperienceValue > MaxExperience) {
+		player._pExperience = MaxExperience;
+	} else if (newExperienceValue > player._pExperience) {
+		player._pExperience = newExperienceValue;
+	} // else experience has overflowed, could either reorder the checks to apply MaxExperience or just let the existing value stay (the current approach)

```

Comment:
Woudn't this work for what we need?
```suggestion
	player._pExperience = clamp(newExperienceValue, player._pExperience, MaxExperience);
```

I would also think it's very unlikely for an overflow to happen here with you having to receive a minimum of `2294967296` xp from one kill in order to do so at this point.

```suggestion
	player._pExperience = std::min(newExperienceValue, MaxExperience);
```

Reply:
yeah, this was being overly cautious. The first suggestion using clamp would be fine, or this can just directly modify _pExperience again and apply the cap after += like the old code did.

Reply:
went with the std::min and no temp variable option :D